spki/src/traits.rs

```diff
@@ -93,3 +96,11 @@ pub trait EncodePublicKey {
         Ok(doc.write_pem_file(path, SubjectPublicKeyInfoRef::PEM_LABEL, line_ending)?)
     }
 }
+
+/// Returns AlgorithmIndentifier associated with the structure. This is mostly useful for Signing
+/// and Validation keys.
+#[cfg(feature = "alloc")]
+pub trait DynAssociatedAlgorithmIdentifier {
+    /// Returns corresponding AlgorithmIndentifier
+    fn algorithm_identifier(&self) -> AlgorithmIdentifierOwned;
+}
```

Comment:
As an alternative to being `alloc`-gated, it could potentially take params generically ala `AlgorithmIdentifier`:

```suggestion
pub trait DynAssociatedAlgorithmIdentifier<Params> {
    /// Returns corresponding AlgorithmIndentifier
    fn algorithm_identifier(&self) -> AlgorithmIdentifier<Params>;
}
```

This is still object safe, and allows the trait to work without `alloc`.

Though there is a convenience argument for the current implementation.

Reply:
This approach would probably simplify writing a blanket impl for a prospective `AssociatedAlgorithmIdentifier` trait that uses an associated constant.

Reply:
I settled on `Owned` because I needed this crate to fill out the `Certificate` fields (which are `Owned` nowadays). I fear that defining it with `<Params>` would result either in duplication or in some ugly type conversions.

Reply:
You'd have to use `DynAssociatedAlgorithmIdentifier<Any>` to get `AlgorithmIdentifierOwned`

Reply:
Okay, let's start with this and I'll try to add a corresponding `AssociatedAlgorithmIdentifier` trait and an blanket impl.

Reply:
@tarcieri I think I'll be able to use associated const implementation to simplify ECDSA support. Internally I developed this as a separate source module, so there were no locality issues. And splitting things to proper source files brings some interesting problems.

Reply:
Okay, if you'd prefer to define that trait that's fine too